### ISSUE TITLE
fft(pack): The overwrite_x parameter can be confusing

### ISSUE THREAD

**Issue by peterbell10** [scipy/scipy#10241]
Related to #10238 

Some people find the argument `overwrite_x` to be confusing (cupy/cupy#1988). The argument allows the implementation to use the input array as scratch space which may or may not have the output written to it. However, some users interpret it as an inplace transform where the output is written back to the input array. 

It has been suggested to remove it entirely but could also be renamed. FFTW calls this `FFTW_DESTROY_INPUT` which makes it clearer that this is not a guarunteed inplace transform/out-parameter.

See also: #2155, #8500

**Comment by larsoner** [comment]
`destroy_x` is less ambiguous. I'd be fine to switching to this name.

**Comment by rgommers** [comment]
Either keyword is actually pretty poor. An argument could be made for using `inplace=False`, and if it's set to True then actually enforcing in-place operations (so raise an error for non-contiguous input). Or `out=`. Or ...

Keeping `overwrite_x` would be the easier route at this point, but we can still decide to change it during the project (the project will be finished before branching 1.4.x).

**Comment by larsoner** [comment]
I'm fine with `out=`. But there are cases where you don't care if the input is destroyed, and it can't actually be done in place (`rfft`), so just having an `out` argument is not sufficient to get the behavior "give me my output with as much computational efficiency as possible, I don't care what you do to the input" which is currently the purpose of the `overwrite_x` args (in `scipy.fftpack` and even `scipy.linalg`). So in my mind it has little to do with in-place-ness, and more to do with computational optimization at the cost of (maybe) trashing your input.

**Comment by rgommers** [comment]
> But there are cases where you don't care if the input is destroyed, and it can't actually be done in place (`rfft`)

Wouldn't that be an argument for simply not giving `rfft` any keyword?

**Comment by larsoner** [comment]
No, because if the `rfft` computation can be optionally sped up by allowing the underlying implementation to destroy the input `x` during the transformation, then we should (optionally) allow it to do so.

**Comment by peterbell10** [comment]
>An argument could be made for using `inplace=False`, and if it's set to True then actually enforcing in-place operations (so raise an error for non-contiguous input). Or `out=`. Or ...

Note that it's not just that it doesn't always overwrite the input, it's also that reading the input array might not give a sensible output. Consider:

```python
x = np.arange(100).reshape((10,10))
y = scipy.fft.fft2(x.T, overwrite_x=True)

assert y.base is x.base
assert y != x
```

>>But there are cases where you don't care if the input is destroyed, and it can't actually be done in place (`rfft`)

>Wouldn't that be an argument for simply not giving rfft any keyword?

It is actually possible, it's just not exploited at the moment.

**Comment by peterbell10** [comment]
I'm having second thoughts about renaming `overwrite_x`. The docs already say
>If True, the contents of x can be destroyed

Which I think is clear enough. Moreover, it adds a needless obstacle in the user's transition from `scipy.fftpack` to `scipy.fft`.

Perhaps a small note is needed to explicitly state that this is not a guarunteed inplace transform.

**Comment by larsoner** [comment]
Fine by me. It `overwrite_x`, even if not the best name, also matches what we call this sort of thing in [scipy.linalg.svd](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.svd.html) and similar functions. So it helps with API consistency.

**Comment by rgommers** [comment]
Okay, after seeing the discussion fine by me too. Can be solved with more explicit docs.

>  also matches what we call this sort of thing in scipy.linalg.svd and similar functions.

good catch

**Comment by peterbell10** [comment]
I've added the following note to #10241

    When ``overwrite_x=True`` is specified, the memory referenced by ``x`` may
    be used by the implementation in any way. This may include reusing the
    memory for the result, but this is in no way guarunteed. You should not
    rely on the contents of ``x`` after the transform as this may change in
    future without warning.

Since we all seem to be in agreement, I'll close the issue.  Feel free to reopen.

**Comment by peterbell10** [comment]
I'm reopening this because @mreineck's continued work on pypocketfft has opened up a new  option. In an update to the [python API][1], he's considering adding an `out` parameter to replace the bool `inplace`. This would make it a lot easier to replace `overwrite_x` with the more conventional `out` parameter as @rgommers suggested.

One issue I can see is that if we remove `overwrite_x`, it could make it harder to port from `scipy.fftpack` to `scipy.fft`. If this is a concern we could possibly support both.

[1]: https://gitlab.mpcdf.mpg.de/mtr/pypocketfft/issues/3#note_41655

**Comment by larsoner** [comment]
To me `overwrite_x` has a different meaning from `out`.

- `overwrite_x` has to do with algorithmic efficiency in cases where you can gain performance by destroying an input (maybe only in some cases, never guaranteed to happen)
- `out` has to do with where you store the result, basically it's there to avoid a `malloc`.

I see uses for both. I think `out` is clearly useful in its own right because it avoids a `malloc`. The questions are:

- Do we actually need `overwrite_x`? In `scipy.linalg` there is often an `overwrite_x` because LAPACK can gain performance by overwriting its inputs sometimes. 
- If `out` is `x`, do things actually work properly? For example if they incrementally write some outputs to `out` and then access those values again, things will break.
- How much should we consider these questions in the context of pocketfft alone, versus the behavior of other FFT libraries (CuPy, PyFFTW) given that we want to make a generic backend?

Since we know some backends will support `overwrite_x` (fftpack) and some will support `out` (pocketfft), perhaps we want both, with the spec:

- `overwrite_x` is more of a hint/optional sort of thing. For ones that don't do anything with it, there is no problem because it's just an efficiency hint.
- `out` must be dealt with properly. Backends can choose to just add `out[:] = native_out[:]` at the end of their code, or if `out is not None` they can choose to raise a `NotImplementedError`.

If we go this route, each backend would need to deal properly with the case that `x` and `out` share a view of the same underlying memory to ensure it does not lead to unexpected results. And also if both `overwrite_x=True` and `out is x` (or share memory) they need to make sure things work.

**Comment by peterbell10** [comment]
I completely agree: in the general case both `out` and `overwrite_x` make sense, even together. Provided the checks are robust enough to prevent bugs of course. I think FFTW can definitely take advantage of both, not so sure about `pyfftw` specifically though.

One other thing to consider is overlapping memory regions, writing to `x` is still dangerous even though `out is not x`.

**Comment by grlee77** [comment]
> If out is x, do things actually work properly?

At least for complex-to-complex FFTs, in-place transforms are possible in both FFTW and CuPy. I think it is possible in pyFFTW when working with FFTW objects directly, but I don't think it can be done via the existing scipy interfaces.

**Comment by grlee77** [comment]
> To me overwrite_x has a different meaning from out.

Agreed. In PyFFTW, setting `overwrite_x` causes the flag `FFTW_DESTROY_INPUT` to be set. This is different from providing the same array as both input and output.

However, in CuPy, if I recall correctly, there is no corresponding "Destroy Input" flag in CUFFT. In that case I think it just attempts to apply the transform in-place when possible, but does not raise an error if this is not possible and a separate output array must be allocated.
